cob_obstacle_distance/src/distance_manager.cpp

```diff
@@ -0,0 +1,386 @@
+/*
+ *****************************************************************
+ * \file
+ *
+ * \note
+ *   Copyright (c) 2015 \n
+ *   Fraunhofer Institute for Manufacturing Engineering
+ *   and Automation (IPA) \n\n
+ *
+ *****************************************************************
+ *
+ * \note
+ *   Project name: care-o-bot
+ * \note
+ *   ROS stack name: cob_control
+ * \note
+ *   ROS package name: cob_obstacle_distance
+ *
+ * \author
+ *   Author: Marco Bezzon, email: Marco.Bezzon@ipa.fraunhofer.de
+ *
+ * \date Date of creation: May, 2015
+ *
+ * \brief
+ *   Implementation of the DistanceManager definitions.
+ ****************************************************************/
+
+
+#include "cob_obstacle_distance/distance_manager.hpp"
+
+#include <string.h>
+#include <stdint.h>
+
+#include <ros/ros.h>
+
+#include <fcl/collision_object.h>
+#include <fcl/collision.h>
+#include <fcl/distance.h>
+#include <fcl/collision_data.h>
+
+#include <std_msgs/Float64.h>
+#include <visualization_msgs/Marker.h>
+
+#include "cob_obstacle_distance/ObstacleDistance.h"
+#include "cob_obstacle_distance/ObstacleDistances.h"
+
+#include <boost/filesystem.hpp>
+#include <boost/filesystem/path.hpp>
+
+#include <ros/package.h>
+
+#define VEC_X 0
+#define VEC_Y 1
+#define VEC_Z 2
+
+DistanceManager::DistanceManager(ros::NodeHandle& nh) : nh_(nh)
+{}
+
+DistanceManager::~DistanceManager()
+{
+}
+
+int DistanceManager::init()
+{
+    KDL::Tree robot_structure;
+    std::string robot_desc_string;
+    nh_.param("/robot_description", robot_desc_string, std::string());
+    if (!kdl_parser::treeFromString(robot_desc_string, robot_structure)){
+        ROS_ERROR("Failed to construct kdl tree from parameter '/robot_description'.");
+        std::string path = ros::package::getPath("cob_obstacle_distance");
+        std::string testdata_path = ros::package::getPath("cob_obstacle_distance") + "/testdata/robot_description.xml";
+        if (!kdl_parser::treeFromFile(testdata_path,
+                                      robot_structure))
+        {
+            ROS_ERROR("Failed to construct kdl tree from test data. ");
+            return -1;
+        }
+
+        ROS_WARN_STREAM("Constructed from test data in \"" << testdata_path << "\".");
+    }
+
+    if(!nh_.getParam("robo_namespace", this->robo_namespace_))
+    {
+        ROS_ERROR("Failed to get parameter \"robo_namespace\".");
+        return -2;
+    }
+
+    if(!nh_.getParam("chain_base_link", this->chain_base_link_))
+    {
+        ROS_ERROR("Failed to get parameter \"chain_base_link\".");
+        return -3;
+    }
+
+    if (!nh_.getParam("chain_tip_link", this->chain_tip_link_))
+    {
+        ROS_ERROR("Failed to get parameter \"chain_tip_link\".");
+        return -4;
+    }
+
+    if(!nh_.getParam("joint_names", this->joints_))
+    {
+        ROS_ERROR("Failed to get parameter \"joint_names\".");
+        return -5;
+    }
```

Comment:
investigate potential of re-using parameters from cob_twist_controller (including the option of "moving them a level up")